traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/ds/LetsEncryptDnsChallengeWatcher.java

```diff
@@ -0,0 +1,173 @@
+    public static final String DEFAULT_LE_DNS_CHALLENGE_URL = "https://${toHostname}/api/1.4/letsencrypt/dnsrecords/";
+    private static final String configFile = "/opt/traffic_router/db/cr-config.json";
+
+    private ConfigHandler configHandler;
+
+    public void setConfigHandler(final ConfigHandler configHandler) {
+        this.configHandler = configHandler;
+    }
+
+    public LetsEncryptDnsChallengeWatcher() {
+        setDatabaseUrl(DEFAULT_LE_DNS_CHALLENGE_URL);
+    }
+
+    @Override
+    @SuppressWarnings("PMD.CyclomaticComplexity")
+    public boolean useData(final String data) {
+        try {
+            final ObjectMapper mapper = new ObjectMapper(new JsonFactory());
+            final HashMap<String, List<LetsEncryptDnsChallenge>> dataMap = mapper.readValue(data, new TypeReference<HashMap<String, List<LetsEncryptDnsChallenge>>>() { });
+            final List<LetsEncryptDnsChallenge> challengeList = dataMap.get("response");
+
+            final JsonNode mostRecentConfig = mapper.readTree(readConfigFile());
+            final ObjectNode deliveryServicesNode = (ObjectNode) JsonUtils.getJsonNode(mostRecentConfig, ConfigHandler.deliveryServicesKey);
+
+
+            challengeList.forEach(challenge -> {
+                final StringBuilder sb = new StringBuilder();
+                sb.append(challenge.getFqdn());
+                if (!challenge.getFqdn().endsWith(".")) {
+                    sb.append('.');
+                }
+                final String challengeDomain = sb.toString();
+                final String fqdn = challengeDomain.substring(0, challengeDomain.length() - 1).replace("_acme-challenge.", "");
+
+                ObjectNode deliveryServiceConfig = null;
+                String dsLabel = "";
+                final StringBuilder nameSb = new StringBuilder();
+                nameSb.append("_acme-challenge");
+                for (final String label : fqdn.split("\\.")) {
+                    deliveryServiceConfig = (ObjectNode) deliveryServicesNode.get(label);
+                    if (deliveryServiceConfig != null) {
+                        dsLabel = label;
+                        break;
+                    } else {
+                        nameSb.append('.');
+                        nameSb.append(label);
+                    }
+                }
+
+                ArrayNode staticDnsEntriesNode;
+
+                if (deliveryServiceConfig.findValue("staticDnsEntries") != null) {
+                    staticDnsEntriesNode = (ArrayNode) deliveryServiceConfig.findValue("staticDnsEntries");
+                } else {
+                    staticDnsEntriesNode = mapper.createArrayNode();
+                }
+
+                if (challenge.getRecord().isEmpty()) {
+                    for (int i = 0; i < staticDnsEntriesNode.size(); i++) {
+                        if (staticDnsEntriesNode.get(i).get("name").equals("_acme-challenge")) {
+                            staticDnsEntriesNode.remove(i);
```

Comment:
mutating the array like this while iterating it is dangerous. I think instead maybe you want to append all of the non-"_acme-challenge" records to a new ArrayNode? not sure, you could possibly also use iterators.

Reply:
done